### PR TITLE
bpx is added as an optional dependency

### LINKED ISSUE #498: [Bug]: Move BPX to an optional dependency on the PyBOP
### Python Version

Python 3.10.12

### Describe the bug

BPX 0.4.0 requires Pydantic<2. This causes conflict when I am trying to use Battery-Data_Toolkit.

### Steps to reproduce the behaviour

When an user wants to use Battery Data Toolkit (BDT) with PyBOP, using any library of BDT will throw an error. For example:

%pip install battery-data-toolkit
%pip install --upgrade pip ipywidgets -q
%pip install pybop -q
%pip install openpyxl
%pip install pandas --upgrade

from batdata.data import BatteryDataset
import plotly.graph_objects as go
import plotly.offline as pyo
import numpy as np

data_path = r"../data/cycler_data/batch_B12A_cell_2.h5" # set a suitable path for a H5 data
dataset = BatteryDataset.from_batdata_hdf(data_path)

---- This will cause an error.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by Dibyendu-IITKGP**
# Description

BPX 0.4.0 requires Pydantic<2. This causes conflict when I am trying to use Battery-Data_Toolkit.

## Issue reference
Fixes #498 

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [ ] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [ ] Examples: A change to existing or additional examples.
- [x] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [ ] Refactoring: Non-functional changes that improve the codebase.
- [ ] Style: Non-functional changes related to code style (formatting, naming, etc).
- [ ] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All unit tests pass: `$ nox -s tests`
- [x] The documentation builds: `$ nox -s doctest`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [ ] Code is well-commented, especially in complex or unclear areas.
- [ ] Added tests that prove my fix is effective or that my feature works.
- [ ] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -54,6 +53,7 @@ dev = [
     "pytest-mock",
     "pytest-xdist",
     "ruff",
+    "bpx>=0.4",
```
Comment (BradyPlanden) [inline pyproject.toml:56]: Let's move `bpx` to it's own optional dependency. Something like:

```python
bpx = [
    "bpx>=0.4",
]
```
with the `all` dependency including it like so:

```python
all = ["pybop[plot]", "pybop[scifem]", "pybop[bpx]"]
```

**Review by BradyPlanden — COMMENTED** [review]
Thanks @Dibyendu-IITKGP, please also include a breaking changelog entry, cheers!

**Comment by Dibyendu-IITKGP** [comment]
The changelog is to be included in **Bug Fixes** right, or **Braking Changes**?

**Comment by agriyakhetarpal** [comment]
Any imports of `bpx` must also be guarded along with this change, so that `import pybop` does not break with having installed PyBOP without optional dependencies (e.g., `pybop[bpx]` and others).

**Comment by agriyakhetarpal** [comment]
Yes, @BradyPlanden, we'll have to document that PyBOP uses optional dependencies to handle certain operations (perhaps similar to https://docs.pybamm.org/en/latest/source/user_guide/installation/#dependencies, though it's a little tricky to manage because one forgets to update this page at times).

Yes, a useful error message is also nice to have; we have some guidelines on this in place, too: https://docs.pybamm.org/en/latest/source/user_guide/contributing.html#managing-optional-dependencies-and-their-imports. It works by handling imports using a programmatic `importlib.import_module("modulename")` pattern, instead of a `import modulename` statement and linking to the docs if the dependency isn't installed. It could also be useful to check out how `pandas` does this nowadays, since our implementation was quite inspired by them back when @arjxn-py implemented it!

File: docs/installation.rst [hunk 2]
```diff
@@ -42,6 +42,28 @@ For those who prefer to install PyBOP from a local clone of the repository or wi
...
+
+.. code-block:: console
+
+    pip install plotly
+
+scikit-fem - This is a dependency for the multi-dimensional pybamm models, and can be installed using:
+
+.. code-block:: console
+
+    pip install scikit-fem[all]
```
Comment (BradyPlanden) [inline docs/installation.rst:57]: same here:

```suggestion
    pip install pybop[scifem]
```

File: docs/installation.rst [hunk 3]
```diff
@@ -42,6 +42,28 @@ For those who prefer to install PyBOP from a local clone of the repository or wi
 
 In editable mode, changes you make to the source code will immediately affect the PyBOP installation without the need for reinstallation.
 
+Optional Packages
+-----------------
+plotly - PyBOP uses plotly>=5.0, and can be installed with:
+
+.. code-block:: console
+
+    pip install plotly
```
Comment (BradyPlanden) [inline docs/installation.rst:51]: Please update these to the optional installation process via PyBOP:

```suggestion
    pip install pybop[plot]
```

File: docs/installation.rst [hunk 4]
```diff
@@ -42,6 +42,28 @@ For those who prefer to install PyBOP from a local clone of the repository or wi
...
+
+.. code-block:: console
+
+    pip install scikit-fem[all]
+
+bpx - To use physics based battery models of Faraday Institution's MSM project Battery Parameter eXchange (BPX) is required:
+
+.. code-block:: console
+
+    pip install bpx
```
Comment (BradyPlanden) [inline docs/installation.rst:63]: and:
```suggestion
    pip install pybop[bpx]
```

**Review by BradyPlanden — COMMENTED** [review]
Thanks for updating this @Dibyendu-IITKGP. A few suggestions on the documentation, otherwise looks good.

File: docs/installation.rst (hunk 4 above)
Comment (Dibyendu-IITKGP) [reply docs/installation.rst:63]: Thanks @BradyPlanden , shall update all these.

File: pyproject.toml [hunk 5]
```diff
@@ -58,7 +57,10 @@ dev = [
 scifem = [
   "scikit-fem>=8.1.0" # scikit-fem is a dependency for the multi-dimensional pybamm models
 ]
-all = ["pybop[plot]", "pybop[scifem]"]
+bpx = [
+    "bpx>=0.4",
+]
+all = ["pybop[plot]", "pybop[scifem]", "pybop[bpx]"]
```
Comment (agriyakhetarpal) [inline pyproject.toml:63]: This can be simplified to

```suggestion
all = ["pybop[plot,scifem,bpx]"]
```

File: docs/installation.rst [hunk 6]
```diff
@@ -42,6 +42,28 @@ For those who prefer to install PyBOP from a local clone of the repository or wi
...
+
+    pip install pybop[scifem]
+
+bpx - To use physics based battery models of Faraday Institution's MSM project Battery Parameter eXchange (BPX) is required:
+
+.. code-block:: console
+
+    pip install pybop[bpx]
+
+For more information of the additional packages, user may have a look `here <https://github.com/pybop-team/PyBOP/blob/develop/pyproject.toml>`_.
```
Comment (agriyakhetarpal) [inline docs/installation.rst:65]: Another line to indicate the use of `pybop[all]` can be added here

**Review by BradyPlanden — APPROVED** [review]
LGTM - thanks @Dibyendu-IITKGP

**Pull request merged**